Add DDL function `pg_get_roledef` for CREATE ROLE

dbgreen opened the pull request:
This commit adds pg_get_roledef function to the catalog in order to show what's the DDL for CREATE ROLE. It respects the current default options and it aligns to the same output that pg_dumpall shows.

references: PG-155

gonzalemario committed:
Add DDL function `pg_get_roledef` for CREATE ROLE

This commit adds pg_get_roledef function to the catalog in order to show
what's the DDL for CREATE ROLE. It respects the current default options
and it aligns to the same output that pg_dumpall shows.

references: PG-155
co-authored-by: Bryan Green <dbryan.green@gmail.com>
co-authored-by: Mario Gonzalez <gonzalemario@gmail.com>

drpalaric commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -549,6 +550,73 @@ static void get_json_table_nested_columns(TableFunc *tf, JsonTablePlan *plan,
+	/* Look up the role in pg_authid */
+	tuple = SearchSysCache1(AUTHOID, ObjectIdGetDatum(roleid));
+	if (!HeapTupleIsValid(tuple)) /* Role not found */
+		PG_RETURN_NULL();
```
would you need an `ereport` for this one? If the role isn't found, then the user didn't create it. That would be a user issue, right? I am thinking that it might be useful for the user to be notified that they didn't create a role, and not that the function just failed for an undetermined reason.

alvherre replied:
What you say is true, *but* this function takes an argument of type regrole, which is going to resolve the role name to OID, and if that role name doesn't exist, then it's going to raise an error, so you won't ever get to this point for a role that doesn't exist, except if it's passed as OID instead of by name.

alvherre replied:
(actually this makes me think that perhaps it would be more consistent to throw errors rather than return null for the cases where the user doesn't exist)

alvherre commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -549,6 +550,73 @@ static void get_json_table_nested_columns(TableFunc *tf, JsonTablePlan *plan,
+
+	roleid = PG_GETARG_OID(0);
+	if (!OidIsValid(roleid))
+		PG_RETURN_NULL();
```
This check for oid 0 is not needed, because you're going to do exactly the same when SearchSysCache1() returns NULL, which it will when you pass this value to it.

alvherre commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -549,6 +550,73 @@ static void get_json_table_nested_columns(TableFunc *tf, JsonTablePlan *plan,
+	/* Extract role attributes */
+	rolname = NameStr(roleform->rolname);
+
+	rolvaliduntil_datum = SysCacheGetAttr(AUTHOID, tuple,
```
Hungarian notation for variable names is overrated. If the variable is already Datum, why do you need to add the suffix "_datum" to its name?  I would remove that, which would make almost every line of code six characters shorter.

drpalaric commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -549,6 +550,73 @@ static void get_json_table_nested_columns(TableFunc *tf, JsonTablePlan *plan,
+
+	roleid = PG_GETARG_OID(0);
+	if (!OidIsValid(roleid))
+		PG_RETURN_NULL();
```
wondering if you need to return either elog or ereport here as well. Seems more of an internal error might be flagged, so elog?

alvherre replied:
I think in this case RETURN_NULL is appropriate, because the user can do `select get_roledef(0)` and it would be weird if that raised an error different from asking for the definition of a role OID that doesn't exist.  Though to be honest I'm not sure that returning NULL rather than raising an error is the right thing to do, especially because the `regrole` type is going to raise an error with a role name that doesn't exist.

alvherre commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -549,6 +550,73 @@ static void get_json_table_nested_columns(TableFunc *tf, JsonTablePlan *plan,
+	appendStringInfoString(&buf, roleform->rolsuper ?
+					" SUPERUSER" : " NOSUPERUSER");
+
+	if (!rolvaliduntil_isnull)
```
You can name this variable "isnull" and it will be identically easy to understand what's going on.  I would put this block right below the SysCacheGetAttr() call though, and move the "char *validuntil" declaration to line 563, so that you don't have to have it in the middle of all the lines that do appendStringInfo.

alvherre commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -549,6 +550,73 @@ static void get_json_table_nested_columns(TableFunc *tf, JsonTablePlan *plan,
+		validuntil_str = DatumGetCString(DirectFunctionCall1(timestamptz_out,
+															 rolvaliduntil_datum));
+		appendStringInfo(&buf, " VALID UNTIL %s", quote_literal_cstr(validuntil_str));
+		pfree(validuntil_str);
```
You don't need to free this value.  It will be freed when the memory context goes away, very soon after this function returns.

alvherre commented on src/test/regress/expected/check_ddl.out:
```diff
@@ -0,0 +1,22 @@
+                              ^
+-- Test role with CONNECTION LIMIT and VALID UNTIL in UTC
+CREATE ROLE regress_role_connlimit_valid_until CONNECTION LIMIT 64 VALID UNTIL '2038-01-18 00:00:00-00';
+set time zone 'utc';
```
you should do RESET TIME ZONE after your new test lines, to avoid messing with the state of the connection.

alvherre commented on doc/src/sgml/func/func-info.sgml:
```diff
@@ -2175,6 +2175,45 @@ SELECT currval(pg_get_serial_sequence('sometable', 'id'));
+         <indexterm>
+          <primary>pg_get_roledef</primary>
+         </indexterm>
+         <function>pg_get_roledef</function> ( <parameter>role</parameter> <type>regclass</type> )
```
`regclass` is wrong; the argument type is `regrole`.

alvherre commented on doc/src/sgml/func/func-info.sgml:
```diff
@@ -2175,6 +2175,45 @@ SELECT currval(pg_get_serial_sequence('sometable', 'id'));
+         Returns the <command>CREATE ROLE</command> command that would recreate the given role.
+         The result is equivalent to the output that <application>pg_dumpall</application>
+         <option>-r</option> would produce for the specified role, except that password
+         information is omitted for security reasons. Returns <literal>NULL</literal> if the
```
I would say "the password information is omitted" without saying why, because the user doesn't really care why. Also, don't repeat below that the password is omitted, because if you repeat it, then it will be repetitive.  Also, note that roles don't have authentication method, they just have a password set or they don't.  It's only pg_hba entries that have authentication method, but that's not relevant to this piece of documentation.

gonzalemario replied:
you sure if you repeat, it will be repetitive? :smiling_face_with_tear:

alvherre commented on doc/src/sgml/func/func-info.sgml:
```diff
@@ -2175,6 +2175,45 @@ SELECT currval(pg_get_serial_sequence('sometable', 'id'));
+        <para>
+         For security reasons, password information is never included in the output, regardless
+         of the role's authentication method. If the specified role does not exist, the function
+         returns <literal>NULL</literal>. The function requires sufficient privileges to view
```
I don't think it's true that the function returns NULL if the role doesn't exist..  At least, the tests seem to contradict you.

gonzalemario replied:
yeah sorry, that was a left over

alvherre commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -549,6 +550,73 @@ static void get_json_table_nested_columns(TableFunc *tf, JsonTablePlan *plan,
+	appendStringInfoString(&buf, roleform->rolbypassrls ?
+					" BYPASSRLS" : " NOBYPASSRLS");
+
+	/* Note: We don't include the password hash for security reasons */
```
If a role has been subject of `ALTER USER SET work_mem` or whatever, I think this function should emit that as well.  Probably also in case of `ALTER USER foo IN DATABASE bar`.

alvherre commented on src/include/catalog/pg_proc.dat:
```diff
@@ -12588,4 +12588,9 @@
   proargnames => '{pid,io_id,io_generation,state,operation,off,length,target,handle_data_len,raw_result,result,target_desc,f_sync,f_localmem,f_buffered}',
   prosrc => 'pg_get_aios' },
 
+{ oid => '9991', descr => 'get CREATE ROLE statement for role by name',
```
actually you can pass a role OID and the function will work just as well, so this comment is slightly bogus.  Also, you're missing a test for the case where you pass a role OID rather than a role name ... I would try something like `SELECT pg_get_roledef(oid) from pg_authid where rolname = 'regress_following_defaults'` or something like that.

alvherre commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -549,6 +550,73 @@ static void get_json_table_nested_columns(TableFunc *tf, JsonTablePlan *plan,
+		pfree(validuntil_str);
+	}
+
+	if (roleform->rolconnlimit >= 0)
```
I would add a one-line comment here "-1 is the default and means no limit"

gonzalemario replied:
Would be better after the `if` or before?

mw2q replied:
My suggestion is if it fits < 80 columns, you can put it on the same line.  Grepping for example suggest either style way can be ok.

gonzalemario replied:
@mw2q thanks! The whole like fits in 79

gonzalemario committed:
WIP: with this the "pg_%" roles are treated as reserved

SELECT pg_get_roledef('pg_database_owner');
ERROR:  role name "pg_database_owner" is reserved

Commit:
This patch implements the pg_get_type_ddl() function, which emits the DDL for CREATE TYPE. It includes functionality comments in the code, as well as tests and documentation.

Author: Phil Alger <phil.alger@enterprisedb.com>

PG-154

dbgreen committed:
added setof capability, replaced dynamic function call with code to handle valid until

dbgreen committed:
updated schedule

dbgreen committed:
Update role_ddl.sql

cleaned up role_ddl.sql

dbgreen committed:
Update role_ddl.out

dbgreen committed:
Merge branch 'postgres:master' into master

dbgreen committed:
Merge branch 'postgres:master' into master

dbgreen committed:
Merge branch 'dev/PG-155-create-role-retail-ddl' of github.com:dbgreen/postgres into dev/PG-155-create-role-retail-ddl

dbgreen committed:
new role_ddl.out

dbgreen committed:
test new functions

dbgreen committed:
new test sql and out

dbgreen committed:
fixed role_ddl tests

dbgreen committed:
fixed role_ddl tests

dbgreen committed:
fixed role_ddl tests

dbgreen committed:
fixed role_ddl tests

dbgreen committed:
fixed role_ddl tests

dbgreen committed:
fixed role_ddl tests
